tools/calyx-writer/lib.rs

```diff
@@ -0,0 +1,1077 @@
+        Marker {
+            element: element.to_string(),
+            id: Some(id.to_string()),
+        }
+    }
+
+    /// Constructs a comment string for the marker at a given `location`. For
+    /// example, `marker.to_string("end")`.
+    pub fn to_string<S: ToString>(&self, location: S) -> String {
+        format!(
+            "// {} {}{}\n",
+            self.element.to_ascii_uppercase(),
+            location.to_string().to_ascii_uppercase(),
+            self.id
+                .as_ref()
+                .map(|id| format!(": {}", id))
+                .unwrap_or_default()
+        )
+    }
+}
+
+/// An element of calyx source code that MUST implement [`CalyxWriter::write`],
+/// can OPTIONALLY implement [`CalyxWriter::marker`], and must NEVER override
+/// [`CalyxWriter::writeln`].
+trait CalyxWriter {
+    fn marker(&self) -> Option<Marker> {
+        None
+    }
+
+    /// Writes this element to `f`. See also [`CalyxWriter::writeln`].
+    fn write(&self, f: &mut IndentFormatter<'_, '_>) -> fmt::Result;
+
+    /// Writes this element followed by a newline and wrapped with markers if
+    /// the element provides them. See also [`CalyxWriter::write`].
+    ///
+    /// Do NOT override this function. See details at [`CalyxWriter`].
+    fn writeln(&self, f: &mut IndentFormatter<'_, '_>) -> fmt::Result {
+        if let Some(marker) = self.marker() {
+            write!(f, "{}", marker.to_string("start"))?;
+            self.write(f)?;
+            f.write_char('\n')?;
+            write!(f, "{}", marker.to_string("end"))?;
+        } else {
+            self.write(f)?;
+            f.write_char('\n')?;
+        }
+        Ok(())
+    }
+}
+
+/// Imports a calyx file from the standard library.
+struct Import {
+    path: String,
+}
+
+impl CalyxWriter for Import {
+    fn write(&self, f: &mut IndentFormatter<'_, '_>) -> fmt::Result {
+        write!(f, "import \"{}\";", self.path)
+    }
+}
+
+/// See `calyx_frontend::Attribute`.
+#[derive(PartialEq, Eq, Clone, Debug)]
+pub struct Attribute {
+    name: String,
+    value: Option<u64>,
+}
+
+impl Attribute {
+    /// Constructs a `calyx_frontend::BoolAttr`, such as `@external`, named
+    /// `name`.
+    pub fn bool<S: ToString>(name: S) -> Self {
+        Self {
+            name: name.to_string(),
+            value: None,
+        }
+    }
+
+    /// Constructs a `calyx_frontend::NumAttr`, such as `@go`, named `name`.    
+    pub fn num<S: ToString>(name: S, value: u64) -> Self {
+        Self {
+            name: name.to_string(),
+            value: Some(value),
+        }
+    }
+}
+
+impl CalyxWriter for Attribute {
+    fn write(&self, f: &mut IndentFormatter<'_, '_>) -> fmt::Result {
+        write!(f, "@{}", self.name)?;
+        if let Some(value) = self.value {
+            write!(f, "({})", value)?;
+        }
+        Ok(())
+    }
+}
+
+/// A list of attributes.
+type Attributes = Vec<Attribute>;
+
+/// Abstracts attribute functionality.
+trait AttributeProvider {
```

Comment:
So this might just be a matter of personal taste but I'm not convinced that this
trait is really doing anything useful? Feels like abstraction for abstraction's
sake, if that makes sense? It's fine to leave it, mostly just thinking aloud

Reply:
Yeah it's probably useless, I can remove it, I just liked that it would enforce a consistent interface for attributes across the things that have them statically (so if I changed the signature it would be a type error unless I changed it everywhere).